train_discreteaction_pendulum.py

```diff
@@ -0,0 +1,242 @@
+import discreteaction_pendulum
+import math
+import random
+import numpy as np
+import matplotlib
+import matplotlib.pyplot as plt
+from collections import namedtuple, deque
+from itertools import count
+
+import torch
+import torch.nn as nn
+import torch.optim as optim
+import torch.nn.functional as F
+
+env = discreteaction_pendulum.Pendulum()
+
+# setup matplotlib
+is_ipython = 'inline' in matplotlib.get_backend()
+if is_ipython:
+    from iPython import display
+
+plt.ion()
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+# create classes for replay memory
+"""tuple for storing a transition"""
+Transition = namedtuple(
+    'Transition', ('state', 'action', 'next_state', 'reward'))
+
+
+class ReplayMemory(object):
+    def __init__(self, capacity) -> None:
+        """Initialize memory"""
+        self.memory = deque([], maxlen=capacity)
+
+    def push(self, *args):
+        """Save a transition into memory"""
+        self.memory.append(Transition(*args))
+
+    def sample(self, batch_size):
+        "Return a random memory"
+        return random.sample(self.memory, batch_size)
+
+    def __len__(self):
+        """Return number of stored memories"""
+        return len(self.memory)
+
+# Create the Q network class
+
+
+class DQN(nn.Module):
+    def __init__(self, n_obs, n_act):
+        super(DQN, self).__init__()
+        self.layer1 = nn.Linear(n_obs, 64)
+        self.layer2 = nn.Linear(64, 64)
+        self.layer3 = nn.Linear(64, n_act)
+
+    def forward(self, x):
+        x = torch.tanh(self.layer1(x))
+        x = torch.tanh(self.layer2(x))
+        return self.layer3(x)
+
+#Training parameters
+MAX_EPS=5000
+MAX_CUDA_EPS=5000
+
+# HYPER PARAMETERS (used as global variables)
+BATCHSIZE = 128  # BATCH SIZE
+GAM = 0.95  # Discount rate
+EPS0 = 0.9  # explorate rate at start
+EPSF = 0.05  # exploration rate at end
+DECAY = 1000  # exploration decay rate
+TAU = 0.005 #target network update rate
+ALF = 1e-4  # learning rate of optimizer
+C = 200     #frequency to reset the target network
+
+actions = range(env.num_actions)
+n_actions = len(actions)
+
+s = env.reset()
+n_obs = len(s)
+
+Q_net = DQN(n_obs, n_actions)  #create a network for the Q function
+target_net = DQN(n_obs, n_actions)  #create a target copy of the Q net
+target_net.load_state_dict(Q_net.state_dict())
+optimizer = optim.AdamW(Q_net.parameters(), lr=ALF, amsgrad=True)
+memory = ReplayMemory(10000)
+
+steps = 0
+
+
+def select_action(state, explore=True):
+    if not torch.is_tensor(state):
+            state = torch.tensor(state, dtype=torch.float32,
+                         device=device).unsqueeze(0)
+            
+    global steps #get steps for exploration decay
+    sample = random.random() #get a float from 0 to 1
+    eps = EPSF+(EPS0-EPSF)*math.exp(-1*steps/DECAY) #calculate epsilon for current step
```

Comment:
Good use of decaying epsilon (not required for hw assignment right?)

Reply:
not required as far as I know